File: flutter/lib/rpc/node.dart
// Copyright (c) 2021 Open Community Project Association https://ocpa.ch
// This software is published under the AGPLv3 license.

/// protobuf handling of protobuf's node RPC messages
///
/// serializes and deserializes protobuf's node messages

import 'dart:typed_data';
import 'package:qaul_app/rpc/protobuf_generated/rpc/qaul_rpc.pb.dart';
import 'protobuf.dart';

// import pre-generated protobuf file
import 'protobuf_generated/node/node.pb.dart';
import 'protobuf_generated/rpc/qaul_rpc.pbenum.dart';

class RpcNode {
  /// decode a binary node protobuf message and process it
  decodeReceivedMessage(List<int> bytes) {
    // decode bytes to message
    Node message = Node.fromBuffer(bytes);

    // send message to the appropriate module
    switch (message.whichMessage()) {
      case Node_Message.info:
        print('RpcNode info message received');
        final nodeInformation = message.ensureInfo();
        final nodeId = nodeInformation.idBase58;
        print('RpcNode node id: $nodeId');
        break;
      default:
        print('UNHANDLED RpcNode protobuf message received');
        break;
    }
  }

  /// encode and send a message
  Future<void> encodeAndSendMessage(Node message) async {
    // send message via qaul RPC
    Rpc rpc = Rpc();
    await rpc.encodeAndSendMessage(Modules.NODE, message.writeToBuffer());
  }

  /// send request node info message
  Future<void> getNodeInfo() async {
    // create message
    Node message = Node();
    message.getNodeInfo = true;

    // send message
    await encodeAndSendMessage(message);
  }
}

File: rust/clients/bridge/src/node.rs
// Copyright (c) 2021 Open Community Project Association https://ocpa.ch
// This software is published under the AGPLv3 license.

//! # Node module functions

use prost::Message;

/// include generated protobuf RPC rust definition file
mod proto { include!("../../../libqaul/src/rpc/protobuf_generated/rust/qaul.rpc.node.rs"); }

/// node module function handling
pub struct Node {}

impl Node {

    /// Process received RPC message
    /// 
    /// Decodes received protobuf encoded binary RPC message
    /// of the node module.
    pub fn rpc(data: Vec<u8>) {
        match proto::Node::decode(&data[..]) {
            Ok(node) => {
                match node.message {
                    Some(proto::node::Message::Info(proto_nodeinformation)) => {
                        // print information
                        println!("Node ID is: {}", proto_nodeinformation.id_base58 );
                        println!("Node Addresses are:");
                        for address in proto_nodeinformation.addresses {
                            println!("    {}", address);
                        }
                    }
                    _ => {},
                }    
            },
            Err(error) => {
                log::error!("{:?}", error);
            },
        }
    }
}